GRC Updates - Draft 2 (#3870)
open-cluster-management/rhacm-docs: governance/config_policy_ctrl.adoc

[#kubernetes-configuration-policy-controller]
= Kubernetes configuration policy controller

> The configuration policy controller can be used to configure any Kubernetes resource and apply security policies across your clusters. The configuration policy is provided in the `policy-templates` field of the policy on the hub cluster, and is propagated to the selected managed clusters by the governance framework. See the xref:../governance/policy_example.adoc#policy-yaml-structure[policy YAML documentation] for more about the hub cluster policy.

Reviewer: ...for more details on...

[#kubernetes-configuration-policy-controller]
= Kubernetes configuration policy controller

> Configuration policy controller can be used to configure any Kubernetes resource and apply security policies across your clusters. The configuration policy is provided in the `policy-templates` field of the policy on the hub cluster, and is propagated to the selected managed clusters by the governance framework. See the xref:../governance/policy_example.adoc#policy-yaml-structure[policy YAML documentation] for more about the hub cluster policy.

Reviewer 2: ```suggestion
The configuration policy controller can be used to configure any Kubernetes resource and apply security policies across your clusters. The configuration policy is provided in the `policy-templates` field of the policy on the hub cluster, and is propagated to the selected managed clusters by the governance framework. See the xref:../governance/policy_example.adoc#policy-yaml-structure[policy YAML documentation] for more about the hub cluster policy.
```

> A Kubernetes object is defined (in whole or in part) in the `object-templates` array in the configuration policy, indicating to the configuration policy controller on the fields to compare with objects on the managed cluster. The configuration policy controller communicates with the local Kubernetes API server to get the list of your configurations that are in your cluster. For more information about CRDs, see https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/[Extend the Kubernetes API with CustomResourceDefinitions].

Reviewer 2: ```suggestion
A Kubernetes object is defined (in whole or in part) in the `object-templates` array in the configuration policy, indicating to the configuration policy controller of the fields to compare with objects on the managed cluster. The configuration policy controller communicates with the local Kubernetes API server to get the list of your configurations that are in your cluster. For more information about CRDs, see https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/[Extend the Kubernetes API with CustomResourceDefinitions].
```

> A Kubernetes object is defined (in whole or in part) in the `object-templates` array in the configuration policy, indicating to the configuration policy controller of the fields to compare with objects on the managed cluster. The configuration policy controller communicates with the local Kubernetes API server to get the list of your configurations that are in your cluster. For more information about CRDs, see https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/[Extend the Kubernetes API with CustomResourceDefinitions].

Reviewer 2: Why are CRDs relevant here?

Author: I left those references in, but I agree they seem a bit too high-level for our docs. 

@dockerymick Do you have thoughts on the CRD links?

Reviewer 3: @dhaiducek not really, just thought it would be helpful to add this reference in case a user was unfamiliar with Kubernetes. I am ok with removing this link reference

> The configuration policy controller is created on the managed cluster during installation. Configuration policy controller supports the `enforce` feature to remediate when the configuration policy is non-compliant. When the `remediationAction` for the configuration policy is set to `enforce`, the controller applies the specified configuration to the target managed cluster.

Reviewer 2: ```suggestion
The configuration policy controller is created on the managed cluster during installation. The configuration policy controller supports the `enforce` feature to remediate when the configuration policy is non-compliant. When the `remediationAction` for the configuration policy is set to `enforce`, the controller applies the specified configuration to the target managed cluster.
```

  namespaceSelector:
    include: ["default"]
    exclude: []
    matchExpressions: []
    matchLabels: {}
  remediationAction: inform
  severity: low
  evaluationInterval:

>     compliant:

Reviewer 2: Should these have values?

Reviewer 3: No, we choose to have an empty policy example so that users could see the framework

| spec.remediationAction
| Required
| Specifies the remediation of your policy. Enter  `inform`.
| `spec.evaluationInterval.noncompliant`
| Optional
| Used to define how often the policy is evaluated when it is in the non-compliant state. Similar to the `evaluationInterval.compliant` parameter, the values must be in the format of a duration which is a sequence of numbers with time unit suffixes. It can also be set to `never` so that the policy is not reevaluated on the noncompliant cluster, unless the policy `spec` is updated.

| spec.remediationAction.severity
| `spec.object-templates`
| Required
| Specifies the severity when the policy is non-compliant. Use the following parameter values: `low`, `medium`, or `high`.
| The array of Kubernetes objects (either fully defined or containing a subset of fields) for the controller to compare with objects on the managed cluster.

> | spec.remediationAction.complianceType

Reviewer 2: It'd be good to also document `metadataComplianceType`:
https://github.com/stolostron/config-policy-controller/blob/09e981ce5a35e6fb500b0c36962383b4cb18fa90/deploy/crds/policy.open-cluster-management.io_configurationpolicies.yaml#L140

> | Used to list expected behavior for roles and other Kubernetes object that must be evaluated or applied to the managed clusters. You must use the following verbs as parameter values:

Reviewer 2: ```suggestion
| Used to define the desired state of the Kubernetes object on the managed clusters. You must use one of the following verbs as the parameter value:
```

| `spec.evaluationInterval.noncompliant`
| Optional
| Used to define how often the policy is evaluated when it is in the non-compliant state. Similar to the `evaluationInterval.compliant` parameter, the values must be in the format of a duration which is a sequence of numbers with time unit suffixes. It can also be set to `never` so that the policy is not reevaluated on the non-compliant cluster, unless the policy `spec` is updated.

| `spec.object-templates`
| Required
| The array of Kubernetes objects (either fully defined or containing a subset of fields) for the controller to compare with objects on the managed cluster.

| `spec.object-templates[].complianceType`
| Required
| Used to define the desired state of the Kubernetes object on the managed clusters. You must use one of the following verbs as the parameter value:

`mustonlyhave`: Indicates that an object must exist with the exact fields and values as defined in the object.

> `musthave`: Indicates an object must exist with the same fields as specified in the objectDefinition. The fields in the template are a subset of what exists in the object. In general, array values are appended. The exception for the array to be patched is when the item contains a `name` key with a value that matches an existing item. If the array should be replaced, use a fully defined `objectDefinition` using the `mustonlyhave` compliance type.

Reviewer: This is more for future reference but we generally try to use active voice where possible (I know it's not always possible).

| `spec.object-templates[].objectDefinition`
| Required
| A Kubernetes object (either fully defined or containing a subset of fields) for the controller to compare with objects on the managed cluster.
|===

See the policy samples that use https://nvd.nist.gov/800-53/Rev4/control/CA-1[NIST Special Publication 800-53 (Rev. 4)], and are supported by {product-title-short} from the https://github.com/stolostron/policy-collection/tree/main/stable/CM-Configuration-Management[`CM-Configuration-Management` folder]. Learn about how policies are applied on your hub cluster, see xref:../governance/policy_sample_intro.adoc#supported-policies[Supported policies] for more details. 

Learn how to create and customize policies, see xref:../governance/manage_policy_overview.adoc#manage-security-policies[Manage security policies]. Refer to xref:../governance/policy_controllers.adoc#policy-controllers[Policy controllers] for more details about controllers.

[#configure-config-policy-controller]
== Configure the configuration policy controller

> You can configure the concurrency of the configuration policy controller per managed cluster to change how many configuration policies it can evaluate at the same time. To change the default value of `2`, set the `policy-evaluation-concurrency` annotation with a non-zero integer within quotes. You can set the value on the `ManagedClusterAddOn` object called `config-policy-controller` in the managed cluster namespace of the hub cluster.

Reviewer: -> You can configure the concurrency of the configuration policy controller for each managed cluster

(change for translation/accessibility)

Author: Updated--thanks!
